Drop ARP requests based on specific scenarios

rveerama1 opened the pull request:
This PR handles special ARP requests. Based on specific scenarios ARP requests will be dropped silently.
Silently discard ARP request if Ethernet address is broadcast and Source IP address is Multicast  adress.
REFERENCES:
    RFC 826: page 4
    RFC 1122: section 3.3.6 {Silently discard link-layer-only b'cast dg's}
    RFC 1812: section 3.3.2, page 34

Networking stack sometimes try to find source address based on destination address. If interface could not find best match then it returns unspecified address (0.0.0.0). Host should not send these packets.

Drop  ARP request with sender IP address set to localhost.

Fixes #11329.
Fixes #11489

rveerama1 committed:
net: arp: Discard ARP request on special scenario

Discard ARP request if Ethernet address is broadcast and
Source IP address is Multicast address.

REFERENCES:
RFC 826: page 4
RFC 1122: section 3.3.6 {Silently discard link-layer-only b'cast dg's}
RFC 1812: section 3.3.2, page 34

ARP related issues fixed.

Fixes #11329

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
net: ipv4: Drop the replies if src address in unspecified

Networking stack sometimes try to find source address based
on destination address. If interface could not find best match
then it returns unspecified address (0.0.0.0). Host should not
send these packets.

IPv4 reply related issues fixed.

Fixes #11329

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 commented:
One of the address combination in TCP connection still fails, I am working on it.

rveerama1 committed:
net: arp: Fix ptype for non arp messages

If network stack found ARP entry for peer address, then it
tries to send pending IP packet. But it always keeps ptype
as ARP only. In this particular scenario it has to be IP
packet.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

pfalcon reviewed (changes requested):
@rveerama1 : I'm sorry, but I would like to see logging explaining why a packet is being dropped. Just like week I spent an hour or so figuring why my packets are dropped silently. Well, that time it turned out to be a logging problem, but can anticipate it happen again with changes like that.

The commit with Eth broadcast address perhaps that as is, but description like "Networking stack sometimes try to find source address based
on destination address. If interface could not find best match
then it returns unspecified address (0.0.0.0)" really makes me wary, please add a logging there (or explain where/how it already happens). Thanks.

rveerama1 commented:
>  I'm sorry, but I would like to see logging explaining why a packet is being dropped

I have added comments in source, but not logging anything.

rveerama1 commented:
> Just like week I spent an hour or so figuring why my packets are dropped silently.

There are many places in if conditions returns some errors. Not every failure if condition has debug logging.

tbursztyka commented:
It would be quite painful to log every place where a packet can be dropped, there are legions of reasons for it. Instead I would revisit ethernet recv function so it would return only at the very end of the function, and if NET_DROP is the verdict it would log it. something like that. So nothing for this PR.

@pfalcon if you could dismiss your review on that

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -642,6 +642,10 @@ struct net_pkt *net_context_create_ipv4(struct net_context *context,
 						  (struct in_addr *)dst);
+		/* If src address is still unspecified, do not create pkt */
+		if (net_ipv4_is_addr_unspecified(src)) {
+			return NULL;
```
Here's where I suggest to add debug logging. Unless you can point to to the fact that net_if_ipv4_select_src_addr() call above already does *clear* logging (like, it hints that if it fails, then some packets can be not sent out, which I guess it doesn't).

pfalcon commented:
> It would be quite painful to log every place where a packet can be dropped

That's why I don't call to do that. What I point it is that there's latest fashion to not log anything at all (like this patch does), and ask to log at least *somewhere*, where the probability of hitting it is (speculatively) high. I added inline comment where I think it's worth to add it, feel free to correct me if it doesn't make sense at that place.

rveerama1 committed:
net: arp: Drop ARP requests from localhost address

If ARP Request with sender IP address set to localhost then drop
the request.

Fixes #11489

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 commented:
@tbursztyka

rveerama1 committed:
net: ipv4: Drop incoming packets with src addr is multicast

As per RFC1112 sec 6.2 "A host group address must never be
placed in the source address field or anywhere in a source
route or record route option of an outgoing IP datagram."

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 commented:
@tbursztyka ping

tbursztyka commented on subsys/net/ip/tcp.c:
```diff
@@ -2358,6 +2358,16 @@ NET_CONN_CB(tcp_syn_rcvd)
 		return NET_DROP;
 	}
 
+#if defined(CONFIG_NET_IPV4)
```
How is this related to #11329, as stated in the commit message?

rveerama1 replied:
ipv4_arp_multicast_icmp.pcap.gz has this packet too. TCP SYN packet with multicast src address.

tbursztyka replied:
Ok, but that's not what #11329 is talking about, there it's only about arp table that should not be updated etc... 

If that syn/mcast is a bug, it's a different one. So do not set Fixes #11329

rveerama1 commented:
@nashif ping

pfalcon commented:
Note that I continue to insist that proper logging is added to the drop cases. We have enough cases of silent dropping already, and when something is logged, it's confusing like hell. So, existing cases should be fixed, like e.g. done in https://github.com/zephyrproject-rtos/zephyr/pull/11662, not more silent-behind-the-user's-back cases added.

tbursztyka commented:
@rveerama1 It kind of make sense so dbg log drop reason actually, can you add such message? (see my comment in PR #11662 to "format" the message )

rveerama1 commented:
Added logging messages.

pfalcon commented on subsys/net/l2/ethernet/arp.c:
```diff
@@ -619,6 +620,16 @@ enum net_verdict net_arp_input(struct net_pkt *pkt)
+		if (memcmp(&eth_hdr->dst, net_eth_broadcast_addr(),
+			   sizeof(struct net_eth_addr)) == 0 &&
+		    net_ipv4_is_addr_mcast(&arp_hdr->src_ipaddr)) {
+			NET_DBG("Drop, eth addr is bcast, src addr is mcast");
```
Thanks for adding logging. But per @tbursztyka's proposal, it should be:

~~~
NET_DBG("DROP: eth addr is bcast, src addr is mcast");
~~~

I think it makes sense to follow that proposal.

rveerama1 replied:
[00:00:03.120,000] <dbg> net_arp.net_arp_input: (0x566bb1c0): DROP: eth addr is bcast, src addr is mcast

Or

[00:00:03.120,000] <dbg> net_arp.net_arp_input: (0x566bb1c0): Drop: eth addr is bcast, src addr is mcast

Or

[00:00:03.120,000] <dbg> net_arp.net_arp_input: (0x566bb1c0): Drop, eth addr is bcast, src addr is mcast

rveerama1 replied:
decide which one make sense?

tbursztyka replied:
first one

pfalcon replied:
Oops, sorry, forgot to link to his post: https://github.com/zephyrproject-rtos/zephyr/pull/11662#discussion_r236543329 . There, he clearly shows `DROP: ` syntax. But can wait for extra confirmation, ping @tbursztyka .

rveerama1 replied:
ok

rveerama1 commented:
Updated.

rveerama1 commented:
Updated.

pfalcon reviewed (approved):
Thanks for updates!

tbursztyka commented:
@pfalcon I am not able to merge anything so let's wait for @nashif or whoever has merge rights.

pfalcon commented:
@tbursztyka:

> @pfalcon I am not able to merge anything so let's wait for @nashif or whoever has merge rights.

Really? I thought you used to have. Anyway, I didn't ask to merge, just checked if you fully understand the changes in https://github.com/zephyrproject-rtos/zephyr/commit/6c872cbd75e1af78f9b8f5144f25f73fd805e6fd of this patch (because I don't from just looking at it).

tbursztyka commented:
@pfalcon I never had merge rights on zephry. About commit 6c872cb, I do understand the change. Actually, while doing my PR #10547 I re-factored ethernet_send (current one is quite a mess), and I did fix that issue by "chance" (Since the code is clearer, is was obvious what to set as ptype there). So current fix is right, but hopefully will switch to a better ethernet_send soon.

pfalcon commented:
@carlescufi, Maybe you could merge this?
